# Drug Resistant GBM: Differential transcript usage analysis
## Mix assembly of short and long reads with StringTie

- First assembly using only the short reads for the sample:
```sh
stringtie -G GRCH38_ANNOTATION.GTF BAM_FILE_FROM_SHORT_READ_ALIGNMENT -o OUTPUT_GTF_NAME
```
- Second assembly using only the long reads for the sample:
```sh
stringtie -L -G GRCH38_ANNOTATION.GTF BAM_FILE_FROM_LONG_READ_ALIGNMENT -o OUTPUT_GTF_NAME
```

- Final assembly using both short and long reads for the sample:
```sh
stringtie --mix -G GRCH38_ANNOTATION.GTF BAM_FILE_FROM_SHORT_READ_ALIGNMENT BAM_FILE_FROM_LONG_READ_ALIGNMENT -o OUTPUT_GTF_NAME
```


Workflow for generating a merged GTF for all StringTie transcripts:

- For each RNA-Seq sample, run StringTie to assemble the read alignments; it is recommended to run StringTie with the -G option if the reference annotation is available.
- Run StringTie with --merge in order to generate a non-redundant set of transcripts observed in any of the RNA-Seq samples assembled previously. The stringtie --merge mode takes as input a list of all the assembled transcripts files (in GTF format) previously obtained for each sample, as well as a reference annotation file (-G option) if available.
- For each RNA-Seq sample, run StringTie using the -B/-b and -e options in order to estimate transcript abundances and generate read coverage tables for expression analysis. The -e option is not required but recommended for this run in order to produce more accurate abundance estimations of the input transcripts. Each StringTie run in this step will take as input the sorted read alignments (BAM file) for the corresponding sample and the -G option with the merged transcripts (GTF file) generated by stringtie --merge. Please note that this is the only case where the -G option is not used with a reference annotation, but with the global, merged set of transcripts as observed across all samples. 

See detailed documentation here: https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual#de


PLEASE FIND THE GENERATED .GTF AND .FA FILES FOR ALL TRANSCRIPTS ALONG WITH THE QUANTIFICATION FILES IN THIS DROPBOX LINK: https://www.dropbox.com/sh/0h9q6c189qgb80o/AACocekoWx4YfKM8W8XdqOLGa?dl=0

## Analysis of differential transcript usage with IsoformSwitchAnalyzeR
Please find in the IsoformSwitchAnalyzeR directory the following:

The R script "isoform_sa_extended_analysis.R" outlines all steps of differential transcript usage analysis. Two types of raw resulting statistics are also provided:
- **isoform_features_not_filtered.csv** : gene- and transcript-level expression statistics for all transcripts
- **bioconsequences.tsv** : biological consequences of the statistically significant transcript switches

Individual splice graphs for all switch genes discovered by IsoforomSwitchAnalyzeR are in the folder "all_diff_splice_plots(not_filtered)".

**NOTE:** The file **"final_list_of_all_switch_genes.txt"** has the final list of switching gene names after the filtering steps outlined in the paper. Other files are raw IsoformSwitchAnalyzeR results.

## Analysis of alternative splicing events with SUPPA2 for filtering results from IsoformSwitchAnalyzeR

 We first generate events for each sample. Note that the following is the example command for sample 42MGBA_1. The analysis should be repeated for all samples:
```sh
python PATH_TO_SUPPA2_DIRECTORY/suppa.py generateEvents -i 42MGBA_1_mix.gtf -o 42MGBA_1.events -e SE SS MX RI FL -f ioe
```
Put all the ioe events in the same file:
```sh
awk '
    FNR==1 && NR!=1 { while (/^<header>/) getline; }
    1 {print}
' *.ioe > all_.events.ioe
```

We then compute psi values per event with the following:
```sh
python PATH_TO_SUPPA2_DIRECTORY/suppa.py psiPerEvent -i all_.events.ioe -e t_data_modified.ctab -o 42MGBA_1_events
```
Finally, we can run the analysis of the differential transcript usage. The .exp files are modified quantification files output by StringTie. Both the raw output from StringTie and the modified versions to fit necessary file format for SUPPA2 are included in this repository. Note that the following is the example command for comparing cell lines 42MGBA and 42MGBA_TMZres. The analysis should be repeated between all 3 cell lines:
```sh
python PATH_TO_SUPPA2_DIRECTORY/suppa.py diffSplice -m empirical -gc -i all_events.ioe -p 42MGBA_events.psi 42MGBA_TMZres_events.psi -e 42MGBA.exp 42MGBA_TMZres.exp  -o 42MGBA_vs_42MGBA_TMZres
```
For a more detailed explanation of the pipeline please see the manual here: https://github.com/comprna/SUPPA
and a tutorial example here: https://github.com/comprna/SUPPA/wiki/SUPPA2-tutorial

All resulting files are in the SUPPA2 directory.

# Mutation Enrichment Analysis Files

Please find all .bed files for mutations discovered on the 3 cell lines in this Dropbox link:
https://www.dropbox.com/sh/0h9q6c189qgb80o/AACocekoWx4YfKM8W8XdqOLGa?dl=0


